src/charm.py

```diff
@@ -43,55 +46,94 @@ def new_func(self, *args, **kwargs):
+            if str(self.model.config["port"]) != relation.data[self.app].get(
+                "port", None
+            ):
+                relation.data[self.app]["port"] = str(self.model.config["port"])
+
+            log.info("Setting relation data: addrs")
+            addrs = []
+            num_units = self.num_units()
+            for i in range(num_units):
+                addrs.append(
+                    UNIT_ADDRESS.format(
+                        self.meta.name, i, self.meta.name, self.model.name
+                    )
+                )
+            if addrs != json.loads(relation.data[self.app].get("addrs", "null")):
+                relation.data[self.app]["addrs"] = json.dumps(addrs)
+
+    @status_catcher
+    def on_alertmanager_changed(self, event):
+        if self.unit.is_leader():
+            self.configure()
+            for relation in self.model.relations["alerting"]:
+                self.update_alerting(relation)
+
+    @status_catcher
+    def on_alertmanager_departed(self, event):
+        if self.unit.is_leader():
+            self.configure()
+            for relation in self.model.relations["alerting"]:
+                self.update_alerting(relation)
 
     @status_catcher
     def on_config_changed(self, _):
+        self.configure()
+        for relation in self.model.relations["alerting"]:
+            self.update_alerting(relation)
+
+    def configure(self):
         """Set Juju / Kubernetes pod spec built from `build_pod_spec()`."""
 
+        self.framework.breakpoint()
+
         if not self.unit.is_leader():
-            log.debug('Unit is not leader. Cannot set pod spec.')
+            log.debug("Unit is not leader. Cannot set pod spec.")
             self.unit.status = ActiveStatus()
             return
 
-        self.framework.breakpoint()
-
         # setting pod spec and associated logging
-        self.unit.status = MaintenanceStatus('Building pod spec.')
-        log.debug('Building pod spec.')
+        self.unit.status = MaintenanceStatus("Building pod spec.")
+        log.debug("Building pod spec.")
 
         pod_spec = self.build_pod_spec()
-        log.debug('Setting pod spec.')
+        log.debug("Setting pod spec.")
         self.model.pod.set_spec(pod_spec)
 
-        for relation in self.model.relations['alerting']:
-            if str(self.model.config['port']) != relation.data[self.app]['port']:
-                log.info('Setting relation data')
-                relation.data[self.app]['port'] = str(self.model.config['port'])
-
         self.unit.status = ActiveStatus()
-        log.debug('Pod spec set successfully.')
+        log.debug("Pod spec set successfully.")
 
     def build_config_file(self):
         """Create the alertmanager config file from self.model.config"""
         if not self.model.config["pagerduty_key"]:
-            raise BlockedStatusError('Missing pagerduty_key config value')
+            raise BlockedStatusError("Missing pagerduty_key config value")
```

Comment:
I am not sure what the requirements are. Is it the case that we do not allow use of alertmanager without pager duty ? Or is it the case we plan to support alternative alerting mechanisms, one of which is pagerduty and that at least one alerting mechanism must be selected ?

Reply:
You need to have some sort of alerting configured. Pagerduty is the only one currently supported as email would require me to write an smtp forwarder charm.